include/deal.II/lac/petsc_ts.templates.h

```diff
@@ -568,20 +568,22 @@ namespace PETScWrappers
           PETSC_ERR_LIB,
           PETSC_ERROR_INITIAL,
           "Failure in ts_ijacobian_with_setup from dealii::PETScWrappers::TimeStepper");
-
       petsc_increment_state_counter(P);
 
       // Handle older versions of PETSc for which we cannot pass a MATSHELL
       // matrix to DMSetMatType. This has been fixed from 3.13 on.
       // What we need to do for older version of PETSc is instead to have
       // a zero matrix with all diagonal entries present.
+      // We use PETSC_MACHINE_EPSILON because some versions of PETSc skip
+      // MatShift if we pass 0.0
       if (user->need_dummy_assemble)
         {
           PetscCall(MatZeroEntries(P));
           PetscCall(MatAssemblyBegin(P, MAT_FINAL_ASSEMBLY));
           PetscCall(MatAssemblyEnd(P, MAT_FINAL_ASSEMBLY));
-          PetscCall(MatShift(P, 0.0));
+          PetscCall(MatShift(P, PETSC_MACHINE_EPSILON));
         }
+      user->need_dummy_assemble = false;
```

Comment:
Is this safe? What if the matrix holds entries of size `1e-52` because we're doing calculations on the nanoscale using SI units? In that case, shifting by `1e-16` is a major change!

Reply:
This matrix is not used at all, since in this case the linear system matrix is mffd and the solve is the user solve from `solve_with_jacobian`

Reply:
Oh, I see. Thanks for pointing it out. If a comment that the matrix isn't actually used were to appear in some future pull request, I would look favorably upon that ;-)

Reply:
I thought `need_dummy_assemble` was clear enough

Reply:
I was apparently not smart enough to understand :-)